Protect exhaustiveness check against missing type equations (#9012,#9019)

garrigue opened the pull request:
Fix #9019 by having `type_pat` always succeed in check mode when trying to type a polymorphic variant pattern and the expected type is a polymorphic variant with an abstract row.

garrigue commented:
As discussed in #9012, I think that this PR fixes an half of the problem, and #9012 the other half.
So I would plead for its inclusion as a safe fix, until we write a more ambitious fix based on the expansion of `row_more`, which may have more consequences.

lpw25 commented:
I'm afraid this replaces one soundness bug with another. Since you cannot unify the variant with the expected type you are not able to get the expected type of the argument of the variant constructor. That is sufficient to cause incompleteness in other ways. For example:

```ocaml
type (_, _) b =
  | A : ([< `A ], 'a) b
  | B : ([< `B of 'a], 'a) b

type _ ty =
  | String_option : string option ty

let f (type x) (type y) (b : (x, y ty) b) (x : x) (y : y) =
  match b, x, y with
  | B, `B String_option, Some s -> print_string s
  | A, `A, _ -> ()

let segfault = f B (`B String_option) None
```

garrigue commented:
Thank you, that's the example I was trying to build of missing expected type information; interesting that my change is the one introducing it...
OK, I'll try a bit harder. Actually, in this case this is not too difficult. Simpler than doing it everywhere.
But still need to properly handle polymorphic variants.

garrigue commented:
OK, not doing anything wasn't a good idea. `unify_pat_types_gadt` won't fail, but is it really sufficient?
At least it solves your case.
Debugging as a game...

garrigue commented:
Also fixed unsoundness found by #9012 by setting the right constructor name before calling disambiguation.
This reverts #9017, as we now need disambiguation for constructors and labels generated by `Parmatch.Conv`.

garrigue commented:
Let me explain the rationale behind this PR.
It intends to fix the unsoundness bug created by the interaction of the incompleteness of GADT unification and the exhaustiveness check.
This comes from the fact some missing equation / type information might be needed in order to type a generated pattern.
Note that is not about GADT indices in the generated pattern, because those are handled by `mcomp` in case of failure. This is about the head type of the generated pattern.
There are two identified problems:
* We might miss an equation on a nominal variant or record type. This means that a constructor might have the wrong type.
* We might only have inferred a partial polymorphic variant type (expansion on locally abstract types in `row_more` is not properly handled).

For the first problem, the nice thing is that since the constructor/label appears inside the type definition, the right constructor can be retrieved from the expected type thanks to disambiguation.
So my fix is very simple: when extracting the constructor/label descriptor from the map, use the name inside the descriptor as `lid`, rather than the meaningless generated name. The disambiguation algorithm does the rest of the work.

For the second problem, the real solution would be to properly handle locally abstract rows, but a temporary fix is to lose in precision of the exhaustiveness check (on the safe side) by calling `unify_pat_types_gadt` instead of `unify_pat_types`. Since we then fallback on `mcomp`, this protects against the incompleteness.
My first fix did the very stupid thing of not unifying at all, which did break an essential invariant of the exhaustiveness check: the type information extracted from the patterns is propagated to the subpatterns, so that no generated pattern has to be typed without type information. By the way, this invariant is also the reason why disambiguation is always going to work on generated constructors/labels, and #9012 should be a valid simplification of the untyping translation in `Parmatch`.
Using `unify_pat_types_gadt` has potentially the side-effect of turning some type variables into existential types. Howver, the above invariant also guarantees that any generated pattern should be typed with a concrete inferred expected type, so those existential types shall have no impact on their typing.

Are there other problematic constructs, outside of constructors, labels and polymorphic variants?
A natural suspect would be extension constructors, as they cannot be retrieved  from their type.
However, an extensible type is by definition never complete, so failing to retrieve a constructor is not enough to make a pattern-matching exhaustive, so this seems ok.
Looking at other cases in `Parmatch.Conv`, `Ppat_var`, `Ppat_any` are by definition ok, `Ppat_constant`, `Ppat_tuple`, `Ppat_array`, `Ppat_lazy` are structural, so that losing equations has no impact on them.

BTW, the invariant of having always a precise enough expected type has an exception: pattern-matching on a polymorphic record field.
```ocaml
type empty = {empty:'a.'a };;
function {empty=None} -> ();;
Warning 8: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
{empty=Some _}
- : empty -> unit = <fun>
```
The irony here is that the warning is actually wrong: since you have no way to produce a value of type `empty`, this match is by definition exhaustive. So we shouldn't have soundness problem with that. However, this might strain some parts of the logic in `Parmatch` or `Matching`.
Of course we can think of more useful polymorphic types:
```ocaml
type el = {el: 'a. 'a list}
type intv = {intv: 'a. [> `Int of int] as 'a}
```
But the point is that there is no way to produce a value corresponding to the polymorphic 'a, so that failing should be OK (I still have to check the interaction with optimizations on or-pattern type checking).

garrigue commented:
The last remark should have no impact on the current `type_pat` or on this PR (we are still able to type generated patterns with no expected type) but might be a problem for #9012.
In particular beware of:
```ocaml
module M : sig type t = A | B end
# function (Some {empty=M.A}) -> ();;
Warning 8: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
(Some {empty=B}|None)
```
If we use an unqualified type, it might become impossible to type `B`, hiding the fact that `None` really needs to be matched.

garrigue commented:
Found yet another problematic case. Not about the pattern itself, but about the expected type. Namely, if we are missing some equation, then we may be left with a flexible abstract type as expected type. But the way we split cases in `Parmatch` means that we might still try to match this type with a concrete pattern. This is clearly untypable, but we should ignore that, because a flexible abstract type could represent any value. I fixed this  by replacing the pattern by `Ppat_any` in that case. I hope this is the end of the story...
I obtained this counter-example by playing with the one from #9012, using a parametric abstract type rather than a redefinition.
Side remark: the current version also applies on private types; this could be refined to expand them with `expand_head_opt`, but then `expected_ty` must be changed too.

garrigue commented:
Actually, after this new counter-example, I have a much better understanding of what is happening.
This seems to be the base case, where the other ones (polymorphic variants and datatypes) are just special cases.

Namely, as @lpw25 explained, the problem is what happens when we split a case of the form
```
_, W1, pat
```
where the two first columns contain GADT witnesses acting on a shared existential variable, and the pattern `pat` has a concrete type depending on this existential variable.
Even if the original pattern is typable thanks to an equation e1 added by `W1`, the splitting of the wild card may introduce new equations that prevent the introduction of e1 in the environment. As a result `pat` may end up being typed with a type different from the original one.

This is something I had completely missed until now. I was always expecting the pattern to be retyped in a context at least as precise as the original one.

Fortunately, thanks to all the witnesses, we know that if we reach `pat` all the equations are valid, which guarantees that the new expected type is still compatible with the original one, i.e. any visible structure has to be the same.

What does it mean in practice: in check mode, we need to be able to type a pattern at any compatible type.
* if this is an abstract type (or private abbreviation in the current PR), just ignore the pattern, as any value could appear
* if this is variant or record type, its identity might be incorrect, but the witnesses guarantee that its constructors/labels must be identical
* if this is a polymorphic variant type, some constructors may be missing from it

My fixes should put us on the safe side in these 3 situations.

gasche commented:
As @trefis and @lpw25 asked in #9012, I think it's a natural question to ask whether it would be possible to add support in the type-checker for arbitrary type equalities, so that we never prevent the introduction of an equation in the environment -- the root of the problem here. You briefly discussed this in https://github.com/ocaml/ocaml/pull/9012#issuecomment-538439858.

In a long-term / "high road" approach, it sounds like this is something like we definitely want to support (both for GADTs and for the module system). In the short term, you propose arrays of fixes that try to barricade this incompleteness, to prevent it from turning into unsoundness. What about the medium term? Is it possible to think of a (possibly invasive) change of implementation that would be backward-compatible, and let us handle arbitrary equations between types in a GADT context?

(Remark: I guess that the problem of any-explosion making patterns untypable can be turned, by writing similar examples in expressions, to a loss of subject reduction -- even in contexts that are reachable by execution, not just in dead code.)

lpw25 commented:
>This is something I had completely missed until now.

Sorry, I should have been clearer in #9019 as this is precisely what I meant by:

> It seems that the current approach to refuting cases relies on the assumption that if a pattern `_, Foo` is typeable then so is `A, Foo` if `A` has the right type. However, that isn't the case [...]

garrigue commented:
@lpw25 Yes, I had read your comment, and understood it. The until now meant "until this issue arised", which is already bad enough. And I still needed the 3rd example to understand the full meaning of this.

@gasche There are two different questions.
What we do currently is not strictly speaking adding equations, but turning non-parametric abstract types into abbreviations.
First, this excludes equations between datatypes, while in theory we could  do that, including in the parametric case, because being compatible means that the datatypes have to be identical (including their type parameters). However, a problem is that  the types of the constructor arguments may change, so this would have a larger impact on typing. Also, in theory this means that we could add equations between types of arguments too. Larger program.
Second, this excludes parametric abstract types. There the problem is much harder, as this would lead to either second-order unification, or handling equations where both sides are non-atomic. This is clearly "high road".

trefis reviewed (changes requested):
Even though there is yet to be a decision about whether the solution is safe / complete / the right approach, I had a brief look at the implementation, and want to comment on the following point:

> So my fix is very simple: when extracting the constructor/label descriptor from the map, use the name inside the descriptor as lid, rather than the meaningless generated name. The disambiguation algorithm does the rest of the work.

You turned on disambiguation on constructors, but not on labels.
So, on your branch, I observe the following:
```
trunk$ cat ../missing_eq.ml 
type ab = { a : int }

module M : sig
  type mab = { a : int }

  type _ t = AB : ab t | MAB : mab t

  val a : mab
  val ab : mab t
end = struct
  type mab = ab = { a : int }

  type _ t = AB : ab t | MAB : mab t

  let a = { a = 42 }
  let ab = AB
end

open M

let f (type x) (t1 : x t) (t2 : x t) (x : x) =
  match t1, t2, x with
  | AB,  AB, { a = _ } -> print_int 1
  | MAB, _,  { a = _ } -> print_int 2
  | _,  AB,  { a = _ } -> print_int 3
  | _, MAB,  { a = _ } -> print_int 4

let () = f M.ab MAB { a = 42 }
trunk$ ./ocamlc -nostdlib -I stdlib -o missing_eq.exe ../missing_eq.ml 
File "../missing_eq.ml", line 25, characters 4-22:
25 |   | _,  AB,  { a = _ } -> print_int 3
         ^^^^^^^^^^^^^^^^^^
Warning 11: this match case is unused.
File "../missing_eq.ml", line 26, characters 4-22:
26 |   | _, MAB,  { a = _ } -> print_int 4
         ^^^^^^^^^^^^^^^^^^
Warning 56: this match case is unreachable.
Consider replacing it with a refutation case '<pat> -> .'
trunk$ ./missing_eq.exe 
4trunk$ 
```

garrigue commented:
That's strange. It's supposed to be enabled in both cases. See `type_label_a_list`.
I'll look into this. Definitely should have added an example with records.

trefis commented:
> It's supposed to be enabled in both cases. See `type_label_a_list`.

I'm looking at it, and it's not calling `disambiguate_lid_a_list` when `labels = Some _`.

garrigue commented:
Indeed, I had fixed the `lid`, but didn't see that disambiguation had to be called separately...
Fixed now. Thanks.

garrigue commented:
Ping.
Or do you think we will have time to discuss that in margin of the next developer meeting?

I have just done a small clean-up, moving the logic of whether the expected type is abstract and compatible with anything to `Ctype`. This now matches correctly what is done in `mcomp`. As a result, there is no longer a strict need to call `unify_pat_types_gadts` for polymorphic variants, as an uninstantiated expected type can type any pattern. However, keeping it gives more precise partiality warnings.

My position: this is a soundness bug, and the fix is now supposed to be safe (i.e. should only result in more generated cases being found typable).

lpw25 commented:
I'll try to find time to look properly before then, but I think that discussing it at the next meeting is a good idea.

lpw25 commented:
Having discussed this with you at the recent meeting and read the code a few times, I'm not keen on a couple of aspects of this patch.

Firstly, the handling of variant constructors and record labels is very strange. It would make sense for parmatch to create fake identifiers and provide a hash table from them to the right label. It would also make sense to leave in the original identifiers and rely on disambiguation to pick the same label again. This code seems to try and do both of those things at once. So it creates fake identifiers and uses them only to lookup the original identifiers and runs disambiguation on those.

Secondly, I think that this code is very unlikely to catch all the cases of incompleteness:
```ocaml
let pdesc =
    if mode = Normal then sp.ppat_desc else
    (* If the expected type is abstract and compatible with any type,
       discard the pattern *)
    match expand_head !env expected_ty with
    | {desc=Tconstr (path, _, _)} when Ctype.is_compatible !env path ->
        Ppat_any
    | _ -> sp.ppat_desc
  in
```
It basically says: "if the expected type is compatible with everything then don't check this pattern". That does not handle any cases where the expected type is not compatible with every type but is compatible with the actual type.

I think that the only reasonable way to fix this part is to make sure that `type_pat` in `Counter_example` mode only ever calls `unify_gadts` not `unify`. That way we know that it will only fail if there is a genuine incompatibility between types. The best way to do this would be to pass a `mode` argument to the `unify_pat_types` function.

garrigue commented:
> Firstly, the handling of variant constructors and record labels is very strange. It would make sense for parmatch to create fake identifiers and provide a hash table from them to the right label. It would also make sense to leave in the original identifiers and rely on disambiguation to pick the same label again. This code seems to try and do both of those things at once. So it creates fake identifiers and uses them only to lookup the original identifiers and runs disambiguation on those.

Indeed, the only role of these fake identifiers is to allow one to look up the definitions.
If some expected type is available, it will be possible to choose the right definition according to the name only. And I suggested rpeatedly that @trefis approach of using only the name was potentially a good alternative. The only advantage of this code is to offer a safe fallback in case no expected type is available. I'm perfectly open to switching to the simpler approach in the future if we can convince ourselves that there is always some type available, but it should probably be done in a separate PR.

> Secondly, I think that this code is very unlikely to catch all the cases of incompleteness:
>  ...
> It basically says: "if the expected type is compatible with everything then don't check this pattern". That does not handle any cases where the expected type is not compatible with every type but is compatible with the actual type.

That case is handled by each individual case:
* for a variant or record type, we allow equivalent types thanks to our use of disambiguation
* for a polymorphic variant type with an abstract we call `unify_gadts`
* extensible variants cannot appear in remaining cases, since `_` always remains in that case
* all other patterns are structural, and as such their head must be equal if it is compatible (but not compatible with everything)

> I think that the only reasonable way to fix this part is to make sure that `type_pat` in `Counter_example` mode only ever calls `unify_gadts` not `unify`. That way we know that it will only fail if there is a genuine incompatibility between types.

I'm not necessarily opposed to doing that. But I would not rely on that alone. For instance, if we were to use that for records with type parameters, we would fail to unify the parameters and propagate their information (actually, the code in the `unique-types` branch would do the right thing, but it is not going to be in 4.10).
Whether we still need `is_compatible` if we do  that is less clear. One would have to check in detail that `expected_ty` is only used through unification and sufficiently late, or in an otherwise safe way.

For now, can we stay on the safe side, and do both?
* keep the code as is
* but switch all unifications to unify_pat_types_gadts in check mode

That way we take no risks.

garrigue committed:
Fix 9019 and some other problems caused by incompleteness

garrigue committed:
in check mode for generated patterns, use only unify_pat_types_gadt

garrigue commented:
I've rebased, compressing all previous commits into one, and adding an extra commit for using `unify_pat_types_gadt` in check mode.
I have a strong belief that the last commit changes nothing (i.e. this will fail in exactly the same cases), but this kind of thing is hard to prove.

trefis commented on testsuite/tests/typing-gadts/pr9019.ml:
```diff
@@ -0,0 +1,236 @@
+module M : sig
+  type mab = A | B
+  type _ t = AB : ab t | MAB : mab t
+  val ab : mab t
+end = struct
+  type mab = ab = A | B
+  type _ t = AB : ab t | MAB : mab t
+  let ab = AB
+end
+[%%expect{|
+type ab = A | B
+module M :
+  sig type mab = A | B type _ t = AB : ab t | MAB : mab t val ab : mab t end
+|}]
+
+open M
+
+let f (type x) (t1 : x t) (t2 : x t) (x : x) =
+  match t1, t2, x with
+  | AB,  AB, A -> 1
+  | MAB, _, A -> 2
+  | _,  AB, B -> 3
+  | _, MAB, B -> 4
+[%%expect{|
+Lines 4-8, characters 2-18:
+4 | ..match t1, t2, x with
+5 |   | AB,  AB, A -> 1
+6 |   | MAB, _, A -> 2
+7 |   | _,  AB, B -> 3
+8 |   | _, MAB, B -> 4
+Warning 8: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+(AB, MAB, A)
+val f : 'x M.t -> 'x M.t -> 'x -> int = <fun>
+|}]
+
+let () = ignore (f M.ab MAB A)
+[%%expect{|
+Exception: Match_failure ("", 4, 2).
+|}]
```
This test is already in  `testsuite/tests/basic-more/robustmatch.ml`

garrigue replied:
I know, I copied it from there.
But we added variations here.

trefis commented on typing/typecore.ml:
```diff
@@ -1431,7 +1446,9 @@ and type_pat_aux
-        unify_head_only loc !env (instance expected_ty) constr;
+        unify_head_only ~gadt loc env (instance expected_ty) constr;
+      (* unify_pat_types ~gadt loc env (instance expected_ty)
+                                      (snd (instance_constructor constr)); *)
```
This is just noise.

garrigue replied:
Yes, I still wonder why this one doesn't work.

trefis commented on typing/typecore.ml:
```diff
@@ -859,7 +854,12 @@ let type_label_a_list ?labels loc closed env type_lbl_a opath lid_a_list k =
+              let lid =
+                Location.mkloc (Longident.Lident lbl.lbl_name) lid.loc in
+              lid,
+              Label.disambiguate () lid env opath (Ok[lbl, fun () -> ()]), a
```
I still find this code somewhat dubious in the cases where we know the expected type (i.e. when opath is `Some ..`).
What happens in that case is:
- we know that the label we want here is `lbl` (that's what parmatch wants us to use)
- if `lbl` is a label for a type that matches the expected type, then use it
- otherwise, if the expected type has a label of the same name: use that

PS: the same remark applies for constructors.

garrigue replied:
In the case we know the expected type, it will be used first.
There is just some extra code that can generate warnings, but they are disabled here.

It is only if there is no expected type that we use the received label. I kind of believe that this cannot happen, and that your approach would be sufficient, but I have no proof of that.

trefis replied:
That's not my point.

I think my point is more: if we're not going to completely trust disambiguation (and I agree that we shouldn't… yet), then I'd rather not use it at all. And I would just always use the received label.

It would just be simpler to think about.

trefis replied:
(Simpler than: "if this then that otherwise fallback with ...")

garrigue replied:
As I explained above, in the current situation the unification might then lose information. We can change that after merging unique-types, then it would be completely equivalent.

garrigue committed:
remove commented-out code

lpw25 commented:
> I'm not necessarily opposed to doing that. But I would not rely on that alone. For instance, if we were to use that for records with type parameters, we would fail to unify the parameters and propagate their information (actually, the code in the unique-types branch would do the right thing, but it is not going to be in 4.10).

The point of using `unify_gadt` everywhere is to make it so that dropping type information does not affect soundness. So with or without the code in your branch it should be safe to only rely on that change.

> For now, can we stay on the safe side, and do both?

Putting in two measures, one that you don't trust and one that I don't trust does not sound safe to me. We need to find a measure that we both agree will be safe. I think that the `unify_gadt`s approach is easier to check the correctness of, so I think we should try to make that one work on its own. Do you have any examples that will fail with just the change to `unify_pat_type`?

For the constructors and labels, I can see why using just disambiguation can potentially fail if we have lost some type information. I can't see what would be wrong with avoiding disambiguation completely in the `Counter_example` case. Maybe the expected type won't be equal to the type of the constructor, but since it now uses `unify_gadt`s that doesn't actually matter.

garrigue commented:
We can try. That essentially amounts to using only my second commit. We do lose in precision, but it could eventually be sound.

garrigue commented:
New simpler version in #9154 

Applying the second half of this PR fixed the testsuite on first try, so this seems correct.

garrigue commented:
Superseded by #9154 
Note however that the code in this PR was based on a more precise analysis of where incompatibilities may arise, and we may want to revisit it later.
